pkg/virt-controller/watch/vm.go

```diff
@@ -664,6 +665,108 @@ func (c *VMController) handleCPUChangeRequest(vm *virtv1.VirtualMachine, vmi *vi
 	return nil
 }
 
+func (c *VMController) VMNodeSelectorPatch(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	var ops []string
+
+	if vm.Spec.Template.Spec.NodeSelector != nil {
+		vmNodeSelector := make(map[string]string)
+		// copy the node selector map
+		for k, v := range vm.Spec.Template.Spec.NodeSelector {
+			vmNodeSelector[k] = v
+		}
+		vmNodeSelectorJson, err := json.Marshal(vmNodeSelector)
+		if err != nil {
+			return err
+		}
+		if vmi.Spec.NodeSelector == nil {
+			ops = append(ops, fmt.Sprintf(`{ "op": "add", "path": "/spec/nodeSelector", "value": %s }`, string(vmNodeSelectorJson)))
+		} else {
+			currentVMINodeSelector, err := json.Marshal(vmi.Spec.NodeSelector)
+			if err != nil {
+				return err
+			}
+			ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/spec/nodeSelector", "value": %s }`, string(currentVMINodeSelector)))
+			ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/spec/nodeSelector", "value": %s }`, string(vmNodeSelectorJson)))
+		}
+
+	} else {
+		ops = append(ops, fmt.Sprintf(`{ "op": "remove", "path": "/spec/nodeSelector" }`))
+	}
+	generatedPatch := controller.GeneratePatchBytes(ops)
+
+	_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, generatedPatch, &v1.PatchOptions{})
+
+	return err
+}
+
+func (c *VMController) VMIAffinityPatch(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	var ops []string
+
+	if vm.Spec.Template.Spec.Affinity != nil {
+		vmAffinity := vm.Spec.Template.Spec.Affinity.DeepCopy()
+		vmAffinityJson, err := json.Marshal(vmAffinity)
+		if err != nil {
+			return err
+		}
+		if vmi.Spec.Affinity == nil {
+			ops = append(ops, fmt.Sprintf(`{ "op": "add", "path": "/spec/affinity", "value": %s }`, string(vmAffinityJson)))
+		} else {
+			currentVMIAffinity, err := json.Marshal(vmi.Spec.Affinity)
+			if err != nil {
+				return err
+			}
+			ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/spec/affinity", "value": %s }`, string(currentVMIAffinity)))
+			ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/spec/affinity", "value": %s }`, string(vmAffinityJson)))
+		}
+
+	} else {
+		ops = append(ops, fmt.Sprintf(`{ "op": "remove", "path": "/spec/affinity" }`))
+	}
+
+	_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, controller.GeneratePatchBytes(ops), &v1.PatchOptions{})
+
+	return err
+}
+
+func (c *VMController) handleAffinityChangeRequest(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	if vmi == nil || vmi.DeletionTimestamp != nil {
+		return nil
+	}
+
+	if vm.Spec.LiveUpdateFeatures == nil || vm.Spec.LiveUpdateFeatures.Affinity == nil {
+		return nil
+	}
+
+	hasNodeSelectorChanged := false
+	if !equality.Semantic.DeepEqual(vm.Spec.Template.Spec.NodeSelector, vmi.Spec.NodeSelector) {
```

Comment:
What do we do about instance types?

Reply:
@xpivarc Thanks for raising this. Could you please remind me what's the trouble with instance types?
Would there be a node selector on the VMI that is not present on the VM?

Reply:
Yes, exactly. The node selector would be missing on VM but will be present on VMI.

Reply:
Yea, I see it on the instanceTypesSpec now. 
Well, honestly, I wouldn't block on this. 

A few thoughts:
 - Since the node selector is on the VirtualuMachineInstanceSpec and it's immutable, we can only `resize` ( switch to a different pod instance type) to modify the affinity.
 - This PR should not affect instance types since the latter doesn't use the `liveUpdateFeatures` at this point.
 - going forward should find a better way to treat affinity within instance types context.

/cc @lyarwood

Reply:
Thanks for the CC @vladikr and apologies for being late to this, I totally missed we had an open PR when we were discussing this last week.

For now if it's set on the instance type but updated as part of this `liveUpdateFeatures` we should see the submission rejected due to the instance type conflicting with the updated `VirtualMachine` [1].

@enp0s3  are you any closer to posting a design doc or PR for `liveUpdateFeatures` support in instance types so we can hammer some of this stuff out in the open?

[1] https://github.com/kubevirt/kubevirt/blob/b66759de11bf88f2f8228c0202bad0d7fa8a7cd1/pkg/instancetype/instancetype.go#L950-L952

Reply:
Yes, but today this PR doesn't affect VMs with instance types.

The VM would need to have `liveUpdateFeaures` enabled before the VMI was created. 
We don't need to block this PR. The instance types enablement can happen at a later stage.

Reply:
@lyarwood ^^

Reply:
Yup was just confirming what the current behaviour will be with this PR.

Reply:
So one can't enable `liveUpdateFeaures` and reference instancetypes as ell?

Reply:
@vladikr Hey! we can reject the addition of `liveUpdateFeautres` to the VM spec like we do with [CPU live update configuration](https://github.com/kubevirt/kubevirt/blob/02d508511fc41f0dccfb78b8d20c3ccfe52b9c93/pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go#L439)